src/group_theory/quotient_group.lean

```diff
@@ -381,4 +381,229 @@ def quotient_quotient_equiv_quotient :
 
 end third_iso_thm
 
+section Zassenhaus
+open subgroup
+
+@[to_additive] lemma normal_subgroup_mul
+  {A B C : subgroup G} (hA : A ≤ C) [hN : (A.subgroup_of C).normal] (hB : B ≤ C) :
+  ((A ⊔ B : subgroup G) : set G) = A * B :=
+begin
+  suffices h : ((A ⊔ B).subgroup_of C : set C) = A.subgroup_of C * B.subgroup_of C,
+  have key : (C.subtype) ⁻¹' (↑A * ↑B) = (C.subtype) ⁻¹' ↑A * (C.subtype) ⁻¹' ↑B,
+  { refine set.preimage_mul_of_injective C.subtype ↑A ↑B subtype.coe_injective _ _;
+    simp only [← monoid_hom.coe_range, subtype_range]; assumption, },
+  have hsub : (A : set G) * B ⊆ C,
+  { rintro _ ⟨p, q, hp, hq, rfl⟩,
+    exact mul_mem _ (hA hp) (hB hq) },
+  apply_fun set.image (C.subtype) at h,
+  simp only [subgroup_of, coe_comap, ← key] at h,
+  simp only [subtype.image_preimage_coe ↑C _, subgroup.coe_subtype,
+    set.inter_eq_self_of_subset_left hsub, inf_of_le_left (sup_le hA hB),
+    set.inter_eq_self_of_subset_left (set_like.coe_subset_coe.mpr (sup_le hA hB))] at h,
+  exact h,
+  simp [subgroup_of_sup A B C hA hB, normal_mul],
+end
+
+@[to_additive] lemma Zassenhaus_subgroup {A' A B' B : subgroup G} (hA : A' ≤ A) (hB : B' ≤ B) :
```

Comment:
I think from this line down there's enough stuff for a new `zassenhaus.lean` file